src/Interpreters/HashJoin.h

```diff
@@ -351,7 +354,15 @@ class HashJoin : public IJoin
     void reuseJoinedData(const HashJoin & join);
 
     RightTableDataPtr getJoinedData() const { return data; }
-    BlocksList releaseJoinedBlocks();
+    BlocksList releaseJoinedBlocks(bool restructure = false);
```

Comment:
I'd rather separate common logic for both hash joins in separate class. In general, it's not good design to tailor one class for usage by another. I guess, the classes which derives from IJoin should basically incapsulate JOIN implementation, that's it.
But such a refactoring is for another PR.